pkg/volumes/do/volumes.go

```diff
@@ -136,54 +155,136 @@ func (a *DOVolumes) findVolumes(filterByRegion bool) ([]*volumes.Volume, error)
+		tagFound := a.matchesTags(&doVolume)
+
+		if tagFound {
+
+			glog.V(2).Infof("Tag Matched for droplet name=%s and volume name=%s", a.dropletName, doVolume.Name)
+
+			var clusterKey string
+			if strings.Contains(doVolume.Name, "etcd-main") {
+				clusterKey = "main"
+			} else if strings.Contains(doVolume.Name, "etcd-events") {
+				clusterKey = "events"
+			} else {
+				clusterKey = ""
+				glog.V(2).Infof("could not determine etcd cluster type for volume: %s", doVolume.Name)
 			}
 
-			if len(doVolume.DropletIDs) == 1 {
-				vol.AttachedTo = strconv.Itoa(doVolume.DropletIDs[0])
-				vol.LocalDevice = getLocalDeviceName(&doVolume)
+			if strings.Contains(a.nameTag, clusterKey) {
+				vol := &volumes.Volume{
+					ProviderID: doVolume.ID,
+					Info: volumes.VolumeInfo{
+						Description: a.ClusterName + "-" + a.nameTag,
+					},
+					MountName: "master-" + doVolume.ID,
+					EtcdName:  doVolume.Name,
+				}
+
+				if len(doVolume.DropletIDs) == 1 {
+					vol.AttachedTo = strconv.Itoa(doVolume.DropletIDs[0])
+					vol.LocalDevice = getLocalDeviceName(&doVolume)
+				}
+
+				glog.V(2).Infof("Found a matching nameTag=%s for cluster key=%s with etcd cluster name = %s; volume name = %s", a.nameTag, clusterKey, a.ClusterName, doVolume.Name)
+
+				myvolumes = append(myvolumes, vol)
 			}
+		}
+	}
+
+	return myvolumes, nil
+}
+
+func (a *DOVolumes) matchesTags(volume *godo.Volume) bool {
+	for k, v := range a.matchTags {
+		// In DO, we don't have tags with key value pairs. So we are storing them separating by a ":"
+		// Create a string like "k:v" and check if they match.
+		doTag := k + ":" + v
 
-			glog.V(2).Infof("Matching DO Volume found with name=%s ID=%s etcdname=%s mountname=%s", doVolume.Name, doVolume.ID, vol.EtcdName, vol.MountName)
+		glog.V(2).Infof("Check for matching volume tag - doTag=%s for volume name = %s", strings.ToUpper(doTag), volume.Name)
 
-			myvolumes = append(myvolumes, vol)
+		volumeTagfound := a.Contains(volume.Tags, doTag)
+		if !volumeTagfound {
+			return false
+		} else {
+			glog.V(2).Infof("Matching tag found for doTag=%s in volume name = %s", strings.ToUpper(doTag), volume.Name)
 		}
 
+		glog.V(2).Infof("Check for matching droplet tag - doTag=%s for droplet name = %s", strings.ToUpper(doTag), a.dropletName)
+
+		// Also check if this tag is seen in the droplet.
+		dropletTagfound := a.Contains(a.dropletTags, doTag)
+		if !dropletTagfound {
+			return false
+		} else {
+			glog.V(2).Infof("Matching tag found for doTag=%s in droplet name = %s", strings.ToUpper(doTag), a.dropletName)
+		}
 	}
 
-	return myvolumes, nil
+	matchingTagCount := 0
+	// now that the tags are matched, check tagkeys if it exists in both volume and droplet tags, and ensure the key values in the volume and droplet tags match.
+	for _, matchTag := range a.matchTagKeys {
+		for _, volumeTag := range volume.Tags {
+			vt := strings.SplitN(volumeTag, ":", -1)
+			vtKey := vt[0]
+			vtValue := vt[1]
```

Comment:
Looks like this will panic if there's a key that isn't of the `key:value` format

Reply:
Thanks, updated this.